miner/worker.go

```diff
@@ -1201,64 +1208,94 @@ func (w *worker) commit(uncles []*types.Header, interval func(), update bool, st
+	maxBundle := simulatedBundle{mevGasPrice: new(big.Int)}
+
 	for _, bundle := range bundles {
+		state, err := w.chain.StateAt(parent.Root())
+		if err != nil {
+			return simulatedBundle{}, err
+		}
+		gasPool := new(core.GasPool).AddGas(header.GasLimit)
 		if len(bundle) == 0 {
 			continue
 		}
-		totalEth, totalGasUsed, err := w.computeBundleGas(bundle, parent, header)
+		simmed, err := w.computeBundleGas(bundle, parent, header, state, gasPool, pendingTxs)
 
 		if err != nil {
 			log.Debug("Error computing gas for a bundle", "error", err)
 			continue
 		}
 
-		mevGasPrice := new(big.Int).Div(totalEth, new(big.Int).SetUint64(totalGasUsed))
-		if mevGasPrice.Cmp(maxBundlePrice) > 0 {
-			maxBundle = bundle
-			maxBundlePrice = mevGasPrice
-			maxTotalEth = totalEth
-			maxTotalGasUsed = totalGasUsed
+		if simmed.mevGasPrice.Cmp(maxBundle.mevGasPrice) > 0 {
+			maxBundle = simmed
 		}
 	}
 
-	return maxBundle, maxBundlePrice, maxTotalEth, maxTotalGasUsed
+	return maxBundle, nil
 }
 
 // Compute the adjusted gas price for a whole bundle
 // Done by calculating all gas spent, adding transfers to the coinbase, and then dividing by gas used
-func (w *worker) computeBundleGas(bundle types.Transactions, parent *types.Block, header *types.Header) (*big.Int, uint64, error) {
-	env, err := w.generateEnv(parent, header)
-	if err != nil {
-		return nil, 0, err
-	}
-
+func (w *worker) computeBundleGas(bundle types.Transactions, parent *types.Block, header *types.Header, state *state.StateDB, gasPool *core.GasPool, pendingTxs map[common.Address]types.Transactions) (simulatedBundle, error) {
 	var totalGasUsed uint64 = 0
 	var tempGasUsed uint64
-	gasFees := new(big.Int)
-
-	coinbaseBalanceBefore := env.state.GetBalance(w.coinbase)
+	totalEth := new(big.Int)
 
 	for _, tx := range bundle {
-		receipt, err := core.ApplyTransaction(w.chainConfig, w.chain, &w.coinbase, env.gasPool, env.state, env.header, tx, &tempGasUsed, *w.chain.GetVMConfig())
+		coinbaseBalanceBefore := state.GetBalance(w.coinbase)
+
+		receipt, err := core.ApplyTransaction(w.chainConfig, w.chain, &w.coinbase, gasPool, state, header, tx, &tempGasUsed, *w.chain.GetVMConfig())
 		if err != nil {
-			return nil, 0, err
+			return simulatedBundle{}, err
 		}
 		if receipt.Status == types.ReceiptStatusFailed {
-			return nil, 0, errors.New("revert")
+			return simulatedBundle{}, errors.New("revert")
 		}
 
 		totalGasUsed += receipt.GasUsed
-		gasFees.Add(gasFees, new(big.Int).Mul(big.NewInt(int64(totalGasUsed)), tx.GasPrice()))
+
+		from, err := types.Sender(w.current.signer, tx)
+		if err != nil {
+			return simulatedBundle{}, err
+		}
+
+		txInPendingPool := false
+		// check if tx is in pending pool
+		if accountTxs, ok := pendingTxs[from]; ok {
+			txNonce := tx.Nonce()
+
+			for _, accountTx := range accountTxs {
```

Comment:
I believe that txs are sorted by nonce here and we can stop searching as soon as nonce is higher than of the transaction that we search?

Reply:
It is actually sorted here, but out of an abundance of caution, iterating the list is safer. The speed savings would be minimal.